src/main/scala/io/iohk/ethereum/jsonrpc/TestService.scala

```diff
@@ -167,39 +168,23 @@ class TestService(
     SetChainParamsResponse().rightNow
   }
 
-  private def storeGenesisAccountCodes(config: BlockchainConfig, accounts: Map[String, GenesisAccount]): Unit = {
-    val genesisBlock = blockchain.getBlockByNumber(0).get
-    val world =
-      blockchain.getWorldStateProxy(0, UInt256.Zero, genesisBlock.header.stateRoot, false, config.ethCompatibleStorage)
-
-    val accountsWithCodes = accounts.filter(pair => pair._2.code.isDefined)
-
-    val worldToPersist = accountsWithCodes.foldLeft(world)((world, addressAccountPair) => {
-      world.saveCode(Address(addressAccountPair._1), addressAccountPair._2.code.get)
-    })
-
-    InMemoryWorldStateProxy.persistState(worldToPersist)
+  private def storeGenesisAccountCodes(accounts: Map[String, GenesisAccount]): Unit = {
+    for {
+      code <- accounts.map(pair => pair._2.code).collect { case Some(code) => code }
+    } {
+      blockchain.storeEvmCode(kec256(code), code).commit()
+    }
   }
 
-  private def storeGenesisAccountStorageData(config: BlockchainConfig, accounts: Map[String, GenesisAccount]): Unit = {
-    val genesisBlock = blockchain.getBlockByNumber(0).get
-    val world =
-      blockchain.getWorldStateProxy(0, UInt256.Zero, genesisBlock.header.stateRoot, false, config.ethCompatibleStorage)
-
-    val accountsWithStorageData = accounts.filter(pair => pair._2.storage.isDefined && pair._2.storage.get.nonEmpty)
-
-    val worldToPersist = accountsWithStorageData.foldLeft(world)((world, addressAccountPair) => {
-      val address = Address(addressAccountPair._1)
-      val emptyStorage = world.getStorage(address)
-      val updatedStorage = addressAccountPair._2.storage.get.foldLeft(emptyStorage) { case (storage, (key, value)) =>
-        storage.store(key, value)
-      }
-      val updatedWorld = world.saveStorage(Address(addressAccountPair._1), updatedStorage)
-      updatedWorld.contractStorages.values.foreach(cs => cs.inner.nodeStorage.persist())
-      updatedWorld
-    })
-
-    InMemoryWorldStateProxy.persistState(worldToPersist)
+  private def storeGenesisAccountStorageData(accounts: Map[String, GenesisAccount]): Unit = {
+    val emptyStorage = domain.EthereumUInt256Mpt.storageMpt(
+      Account.EmptyStorageRootHash,
+      blockchain.getStateStorage.getBackingStorage(0)
+    )
+    val storagesToPersist = accounts.map(pair => pair._2.storage).collect { case Some(map) if map.nonEmpty => map }
+    for { storage <- storagesToPersist } {
+      storage.foldLeft(emptyStorage) { case (storage, (key, value)) => storage.put(key, value) }
```

Comment:
Why `foldLeft` and not another `for` loop or a `foreach`? Since it's only about the side-effect right.

Reply:
It's about the side effect, but it's also important to get the return of `put`. Because it stores the new MPT in the database but does not remove the previous MPT, it just give back the new MPT corresponding to the modification (that's why I can reuse `emptyStorage` each time).

But this make me think that it's not really efficient as it creates intermediate MPT each time. I can use `update` which allows to add several keys at once. I will change this